File: Rust/greedy.rs
// rustc -C opt-level=3 -C target-cpu=core2 greedy.rs

use std::old_io::File;
use std::str::FromStr;
use std::f64;
use std::f64::consts::PI;
use std::num::Float;

static EARTH_RADIUS_M:f64 = 6367450.0;
static CONVERT_RAD:f64 = PI / 180.0;

#[derive(Clone)]
struct Activity {
    id: String,
    lat: f64,
    lon: f64,
}

impl Activity {
    fn new(id: &String, lat: f64, lon: f64) -> Activity {
        Activity { id: id.clone(), lat: lat, lon: lon }
    }
}

#[derive(Clone)]
struct Resource {
    id: String,
    lat: f64,
    lon: f64,
}

impl Resource {
    pub fn new(id: &String, lat: f64, lon: f64) -> Resource {
        Resource { id: id.clone(), lat: lat, lon: lon }
    }
}

#[derive(Clone)]
struct Allocation {
    rid: String,
    aid: String,
    dist: f64
}

impl Allocation {
    pub fn new(rid: &String, aid: &String, dist: f64) -> Allocation {
        Allocation { rid: rid.clone(), aid: aid.clone(), dist: dist }
    }
}

#[derive(Clone)]
struct SchemaData {
    activity: Vec<Activity>,
    resource: Vec<Resource>,
}

impl SchemaData {
    pub fn new() -> SchemaData {
        SchemaData { 
            activity: Vec::new(), 
            resource: Vec::new()
        }
    }
}

fn build(lines: Vec<&str>) -> SchemaData {
    let mut sd = SchemaData::new();
    for x in lines.iter() {
        let split = x.split_str(",").collect::<Vec<&str>>();
        let id:&str = split[0];
        let lat: f64 = FromStr::from_str(split[1]).unwrap();
        let lng: f64 = FromStr::from_str(split[2]).unwrap();
        if split.len() == 3 {
            sd.resource.push(Resource::new(&String::from_str(id), lat, lng));
        } else if split.len() == 4 {
            sd.activity.push(Activity::new(&String::from_str(id), lat, lng));
        }
    }
    sd
}

fn distance_between_points(lat1:f64, lon1:f64, lat2:f64, lon2:f64) -> f64 {
    let start_lat_in_rad: f64 = lat1 * CONVERT_RAD;
    let start_lon_in_rad: f64 = lon1 * CONVERT_RAD;
    let end_lat_in_rad: f64 = lat2 * CONVERT_RAD;
    let end_lon_in_rad: f64 = lon2 * CONVERT_RAD;
    let longitude: f64 = end_lon_in_rad - start_lon_in_rad;
    let latitude: f64 = end_lat_in_rad - start_lat_in_rad;
    let sin_half_lat: f64 = Float::sin(latitude * 0.5);
    let sin_half_lon: f64 = Float::sin(longitude * 0.5);
    let a: f64 = sin_half_lat * sin_half_lat + Float::cos(start_lat_in_rad) * Float::cos(end_lat_in_rad) * sin_half_lon * sin_half_lon;
    let c: f64 = Float::atan2(Float::sqrt(a), Float::sqrt(1.0 - a));
    EARTH_RADIUS_M * (c + c)
}

fn schedule_ind(act: &Vec<Activity>, id: &String, lat: f64, lon: f64) -> (Allocation, usize) {
    let mut lowest: f64 = f64::INFINITY;
    let mut lowest_id = &String::new();
    let mut i = 0;
    let mut j = 0;
    for act in act.iter() {
        let dist = distance_between_points(lat, lon, act.lat, act.lon); 
        if dist < lowest {
            lowest = dist;
            lowest_id = &act.id;
            j = i;
        }
        i += 1;
    }
    let a = Allocation::new(id, &lowest_id, lowest);
    (a, j)
}

fn schedule_resources(sd: &mut SchemaData) -> f64 {
    let mut allocation: Vec<Allocation> = Vec::new();
    for res in sd.resource.iter() {
        for _ in 0..50 {
            let (a, j) = schedule_ind(&sd.activity, &res.id, res.lat, res.lon);
            sd.activity.remove(j);
            allocation.push(a);
        }
    }

    // Sum it
    let mut sum: f64 = 0.0;
    for all in allocation.iter() {
        sum += all.dist;
    }
    sum
}

fn main() {
    // Load lines
    let mut file = File::open(&Path::new("/Users/daryl/Development/Projects/FunctionalComparison/Data/DataSPIF.csv"));
    let contents = file.read_to_end().unwrap();
    let filestr = String::from_utf8(contents).unwrap();
    let lines: Vec<&str> = filestr.lines().collect();

    // Build
    let sd = build(lines);

    // Schedule
    for i in 0..100 {
        let mut sd2 = sd.clone();
        let sum = schedule_resources(&mut sd2);
        println!("{}: {}", i, sum);
    }
}

File: Java/Greedy.java
// javac Greedy.java -Xdiags:verbose
// time java -server -Xms1024m -XX:+UseG1GC `basename Greedy .java`

import java.util.*;
import java.nio.file.*;
import java.nio.file.Files.*;
import java.util.stream.Collectors;
import java.io.IOException;

class Greedy {
    final static double earthRadiusM = 6367450;
    final static double convert2Rad = Math.PI / 180.0;
    final static double convert2Deg = 180.0 / Math.PI;
    final static double seconds_per_metre = 0.0559234073;

    static class Activity 
    {
        public String id;
        public double lat;
        public double lon;
    }

    static class Resource 
    {
        public String id;
        public double lat;
        public double lon;
    }

    static class Allocation
    {
        public String resourceId;
        public String activityId;
        public double distance;
    }

    static class SchemaData
    {
        public Map<String, Activity> activity = new HashMap<>();
        public List<Resource> resource = new LinkedList<>();
        public List<Allocation> allocation = new LinkedList<>();
    }

    private static double distanceBetweenPointsLatLong(double lat1, double lon1, double lat2, double lon2) {
        double dStartLatInRad = lat1 * convert2Rad;
        double dStartLongInRad = lon1 * convert2Rad;
        double dEndLatInRad = lat2 * convert2Rad;
        double dEndLongInRad = lon2 * convert2Rad;
        double dLongitude = dEndLongInRad - dStartLongInRad;
        double dLatitude = dEndLatInRad - dStartLatInRad;
        double dSinHalfLatitude = Math.sin(dLatitude * 0.5);
        double dSinHalfLongitude = Math.sin(dLongitude * 0.5);
        double a = dSinHalfLatitude*dSinHalfLatitude + Math.cos(dStartLatInRad)*Math.cos(dEndLatInRad)*dSinHalfLongitude*dSinHalfLongitude;
        double c = Math.atan2(Math.sqrt(a), Math.sqrt(1.0-a));
        return earthRadiusM * (c + c);
    }

    private static SchemaData Build(List<String> lines) {
        SchemaData sd = new SchemaData();
        for (String line : lines) {
            String[] split = line.split(",");
            double lat = Double.parseDouble(split[1]);
            double lon = Double.parseDouble(split[2]);
            switch (split.length) {
                case 3:
                    Resource r = new Resource();
                    r.id = split[0];
                    r.lat = lat;
                    r.lon = lon;
                    sd.resource.add(r);
                    break;
                case 4:
                    Activity a = new Activity();
                    a.id = split[0];
                    a.lat = lat;
                    a.lon = lon;
                    sd.activity.put(a.id, a);
                    break;
            }
        }
        return sd;
    }

    private static void scheduleResources(SchemaData sd) {
        for (Resource res : sd.resource) {
            for (int c = 0; c < 50; c++) {
                double lowest = Double.POSITIVE_INFINITY;
                String lowestid = "";
                for (Activity act : sd.activity.values()) {
                    double dist = distanceBetweenPointsLatLong(res.lat, res.lon, act.lat, act.lon);
                    if (dist < lowest) {
                        lowest = dist;
                        lowestid = act.id;
                    }
                }
                Allocation a = new Allocation();
                a.resourceId = res.id;
                a.activityId = lowestid;
                a.distance = lowest;
                sd.allocation.add(a);
                sd.activity.remove(lowestid);
            }
        }
    }

    public static void main(String [] args) throws IOException
    {
        List<String> teams = java.nio.file.Files.lines(Paths.get("../Data/","DataSPIF.csv")).collect(Collectors.toList());  
        SchemaData sd = Build(teams);
        for (int i = 0; i < 500; i++) {
            SchemaData sdLoop = new SchemaData();
            sdLoop.activity = new HashMap<>(sd.activity);
            sdLoop.resource = new LinkedList<>(sd.resource);
            scheduleResources(sdLoop);
            double sum = 0;
            for (Allocation a : sdLoop.allocation) {
                sum += a.distance;
            }
            System.out.println(i + " : " + sum);
        }
    }
}
